File: python/tree_constructor.py
def tree_constructor(strArr) -> bool: 
    parents = {}
    children = {}

    for pair in strArr:
        child, parent = pair.replace("(", "").replace(")", "").split(",")

        if parent not in parents:
            parents[parent] = []
        
        parents[parent].append(child)

        # Check if the parent has more than 2 children (invalid binary tree)
        if len(parents[parent]) > 2:
          return False
        
        # Check if the child already has a parent (invalid binary tree)
        if child in children:
          return False
        
        children[child] = parent

    rootCount = len(list(filter(lambda parent: all(child not in parents for child in parents[parent]), parents)))
    # for parent in parents:
    #    if parent not in children:
    #       rootCount += 1 

    # There should be exactly one root for a valid binary tree
    return rootCount == 1

File: rust/src/tree_constructor.rs
use std::collections::HashMap;

pub fn get(str_arr: Vec<&str>) -> bool {
    let mut parents: HashMap<String, Vec<String>> = HashMap::new();
    let mut children: HashMap<String, String> = HashMap::new();

    for &pair in &str_arr {
        // let cleaned_pair = pair.trim_matches(|c| c == '(' || c == ')');
        let cleaned_pair = pair.replace(['(', ')'], "");
        let nodes: Vec<_> = cleaned_pair.split(',').collect();
        let child = nodes[0].to_owned();
        let parent = nodes[1].to_owned();

        // Insert the child into the parent's children list
        parents
            .entry(parent.clone())
            .or_default()
            .push(child.clone());

        // If a parent has more than two children, it's not a valid binary tree
        if parents[&parent].len() > 2 {
            return false;
        }

        // If a child already has a parent, it's not a valid binary tree
        if children.contains_key(&child) {
            return false;
        }

        // Register the parent of the child
        children.insert(child, parent);
    }

    let root_count = parents
        .keys()
        .filter(|&parent| !children.contains_key(parent))
        .count();
    root_count == 1
}
